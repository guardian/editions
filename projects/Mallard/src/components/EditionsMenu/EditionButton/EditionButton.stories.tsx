import React from 'react'
import { storiesOf } from '@storybook/react-native'
import { withKnobs } from '@storybook/addon-knobs'
import { EditionButton } from './EditionButton'

const specialProps = {
    imageUri:
        'https://media.guim.co.uk/49cebb0db4a3e4d26d7d190da7be4a2e9bd7534f/0_0_103_158/103.png',
    expiry: new Date(),
    onPress: () => {},
    title: 'Food Monthly',
    subTitle: 'The best food of the month',
    titleColor: '#FEEEF7',
    isSpecial: true,
}

storiesOf('EditionsMenu', module)
    .addDecorator(withKnobs)
    .add('RegionButton - default', () => (
        <EditionButton
            onPress={() => {}}
            title="Daily Edition"
            subTitle={'Published every day by 6am (GMT)'}
        />
    ))
    .add('RegionButton - selected', () => (
        <EditionButton
            selected
            onPress={() => {}}
            title={'Daily Edition'}
            subTitle={'Published every day by 6am (GMT)'}
        />
    ))

    .add('SpecialEditionButton - default', () => (
        <EditionButton {...specialProps} />
    ))
    .add('SpecialEditionButton - selected', () => (
        <EditionButton selected {...specialProps} />
    ))
