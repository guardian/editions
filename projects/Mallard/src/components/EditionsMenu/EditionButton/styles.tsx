import { brand, neutral } from '@guardian/src-foundations'
import { StyleSheet } from 'react-native'
import { families } from 'src/theme/typography'

export const EDITIONS_MENU_TEXT_LEFT_PADDING = 75
const imageWidth = 75
const imageHeight = 150

const styles = (selected: boolean, special: boolean, titleColor: string) => {
    return StyleSheet.create({
        container: {
            borderColor: selected ? brand[400] : neutral[86],
            borderWidth: selected ? 4 : 1,
            borderRadius: 3,
            paddingBottom: special ? 8 : 30,
            paddingTop: special ? 14 : 10,
            flexDirection: 'row',
        },
        imageContainer: {
            width: EDITIONS_MENU_TEXT_LEFT_PADDING,
            marginRight: 13,
        },
        textContainer: {
            flex: 1,
        },
        title: {
            fontSize: special ? 32 : 20,
            lineHeight: special ? 34 : 20,
            fontFamily: special
                ? families.headline.medium
                : families.titlepiece.regular,
            marginBottom: 5,
            color: titleColor,
        },
        subTitle: {
            flexWrap: 'wrap',
            paddingRight: 20,
            color: neutral[7],
        },
        expiry: {
            flexWrap: 'wrap',
            paddingRight: 20,
            color: neutral[46],
        },
        image: {
            width: imageWidth,
            height: imageHeight,
        },
    })
}

export { styles }
