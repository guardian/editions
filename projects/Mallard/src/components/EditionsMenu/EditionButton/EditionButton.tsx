import { brand } from '@guardian/src-foundations/palette';
import React, { useState } from 'react';
import { TouchableWithoutFeedback, View } from 'react-native';
import Image from 'react-native-fast-image';
import { TitlepieceText, UiExplainerCopy } from 'src/components/styled-text';
import { localDate } from 'src/helpers/date';
import { styles } from './styles';

interface Props {
	expiry?: Date;
	imageUri?: string;
	onPress: () => void;
	selected?: boolean;
	subTitle: string;
	title: string;
	titleColor?: string;
	isSpecial?: boolean;
}

const EditionButton: React.FC<Props> = ({
	title,
	subTitle,
	imageUri,
	expiry,
	titleColor = brand[400],
	isSpecial = false,
	selected,
	onPress,
}) => {
	const [pressed, setPressed] = useState(false);
	const defaultStyles = styles(selected ?? pressed, isSpecial, titleColor);

	return (
		<TouchableWithoutFeedback
			accessibilityRole="button"
			accessibilityLabel={`${title} - Edition Button`}
			onPressIn={() => setPressed(true)}
			onPressOut={() => setPressed(false)}
			onPress={onPress}
		>
			<View style={defaultStyles.container}>
				<View style={defaultStyles.imageContainer}>
					{imageUri && (
						<Image
							key={imageUri}
							style={defaultStyles.image}
							source={{ uri: imageUri }}
						/>
					)}
				</View>
				<View style={defaultStyles.textContainer}>
					<TitlepieceText style={defaultStyles.title}>
						{title}
					</TitlepieceText>
					<UiExplainerCopy style={defaultStyles.subTitle}>
						{subTitle}
					</UiExplainerCopy>
					{expiry && (
						<UiExplainerCopy style={defaultStyles.expiry}>
							Available until {localDate(expiry)}
						</UiExplainerCopy>
					)}
				</View>
			</View>
		</TouchableWithoutFeedback>
	);
};

export { EditionButton };
