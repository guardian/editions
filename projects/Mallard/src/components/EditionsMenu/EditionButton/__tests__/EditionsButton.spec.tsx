import React from 'react';
import type { ReactTestRendererJSON } from 'react-test-renderer';
import TestRenderer from 'react-test-renderer';
import { EditionButton } from '../EditionButton';

const props = {
	onPress: () => {},
	title: 'Australia Weekend',
	subTitle: 'Published every Saturday by 6am (AEST)',
};

const specialProps = {
	expiry: new Date(98, 1),
	image: {
		source: 'media',
		path: '/path/to/image',
	},
	onPress: () => {},
	title: `Food
Monthly`,
	subTitle: 'Store cupboard special: 20 quick and easy lockdown suppers',
	titleColor: '#FEEEF7',
	buttonImageUri:
		'https://media.guim.co.uk/49cebb0db4a3e4d26d7d190da7be4a2e9bd7534f/0_0_103_158/103.png',
};

jest.mock('src/components/front/image-resource', () => ({
	ImageResource: () => 'ImageResource',
}));

describe('RegionButton', () => {
	it('should display a default RegionButton with correct styling', () => {
		const component: ReactTestRendererJSON | null = TestRenderer.create(
			<EditionButton {...props} />,
		).toJSON();
		expect(component).toMatchSnapshot();
	});
	it('should display a selected RegionButton with correct styling', () => {
		const component: ReactTestRendererJSON | null = TestRenderer.create(
			<EditionButton selected={true} {...props} />,
		).toJSON();
		expect(component).toMatchSnapshot();
	});
});

describe('SpecialEditionButton', () => {
	it('should display a default SpecialEditionButton with correct styling', () => {
		const component: ReactTestRendererJSON | null = TestRenderer.create(
			<EditionButton isSpecial {...specialProps} />,
		).toJSON();
		expect(component).toMatchSnapshot();
	});
	it('should display a selected SpecialEditionButton despite correct styling', () => {
		const component: ReactTestRendererJSON | null = TestRenderer.create(
			<EditionButton isSpecial selected {...specialProps} />,
		).toJSON();
		expect(component).toMatchSnapshot();
	});
});
