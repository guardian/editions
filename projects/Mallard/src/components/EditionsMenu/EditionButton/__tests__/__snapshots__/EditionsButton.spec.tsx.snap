// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegionButton should display a default RegionButton with correct styling 1`] = `
<View
  accessibilityLabel="Australia Weekend - Edition Button"
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "#DCDCDC",
      "borderRadius": 3,
      "borderWidth": 1,
      "flexDirection": "row",
      "paddingBottom": 32,
      "paddingTop": 10,
    }
  }
>
  <View
    style={
      Object {
        "marginRight": 13,
        "width": 75,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "flexShrink": 0,
            "fontFamily": "GTGuardianTitlepiece-Bold",
            "fontSize": 30,
            "lineHeight": 33,
          },
          Object {
            "color": "#052962",
            "fontFamily": "GTGuardianTitlepiece-Bold",
            "fontSize": 20,
            "lineHeight": 20,
            "marginBottom": 5,
          },
        ]
      }
    >
      Australia Weekend
    </Text>
    <Text
      style={
        Array [
          Object {
            "flexShrink": 0,
            "fontFamily": "GuardianTextSans-Regular",
            "fontSize": 15,
            "lineHeight": 18,
          },
          Object {
            "color": "#333333",
          },
          Object {
            "color": "#121212",
            "flexWrap": "wrap",
            "paddingRight": 20,
          },
        ]
      }
    >
      Published every Saturday by 6am (AEST)
    </Text>
  </View>
</View>
`;

exports[`RegionButton should display a selected RegionButton with correct styling 1`] = `
<View
  accessibilityLabel="Australia Weekend - Edition Button"
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "#052962",
      "borderRadius": 3,
      "borderWidth": 4,
      "flexDirection": "row",
      "paddingBottom": 32,
      "paddingTop": 10,
    }
  }
>
  <View
    style={
      Object {
        "marginRight": 13,
        "width": 75,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "flexShrink": 0,
            "fontFamily": "GTGuardianTitlepiece-Bold",
            "fontSize": 30,
            "lineHeight": 33,
          },
          Object {
            "color": "#052962",
            "fontFamily": "GTGuardianTitlepiece-Bold",
            "fontSize": 20,
            "lineHeight": 20,
            "marginBottom": 5,
          },
        ]
      }
    >
      Australia Weekend
    </Text>
    <Text
      style={
        Array [
          Object {
            "flexShrink": 0,
            "fontFamily": "GuardianTextSans-Regular",
            "fontSize": 15,
            "lineHeight": 18,
          },
          Object {
            "color": "#333333",
          },
          Object {
            "color": "#121212",
            "flexWrap": "wrap",
            "paddingRight": 20,
          },
        ]
      }
    >
      Published every Saturday by 6am (AEST)
    </Text>
  </View>
</View>
`;

exports[`SpecialEditionButton should display a default SpecialEditionButton with correct styling 1`] = `
<View
  accessibilityLabel="Food
Monthly - Edition Button"
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "#DCDCDC",
      "borderRadius": 3,
      "borderWidth": 1,
      "flexDirection": "row",
      "paddingBottom": 32,
      "paddingTop": 14,
    }
  }
>
  <View
    style={
      Object {
        "marginRight": 13,
        "width": 75,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "flexShrink": 0,
            "fontFamily": "GTGuardianTitlepiece-Bold",
            "fontSize": 30,
            "lineHeight": 33,
          },
          Object {
            "color": "#FEEEF7",
            "fontFamily": "GHGuardianHeadline-Medium",
            "fontSize": 32,
            "lineHeight": 34,
            "marginBottom": 5,
          },
        ]
      }
    >
      Food
Monthly
    </Text>
    <Text
      style={
        Array [
          Object {
            "flexShrink": 0,
            "fontFamily": "GuardianTextSans-Regular",
            "fontSize": 15,
            "lineHeight": 18,
          },
          Object {
            "color": "#333333",
          },
          Object {
            "color": "#121212",
            "flexWrap": "wrap",
            "paddingRight": 20,
          },
        ]
      }
    >
      Store cupboard special: 20 quick and easy lockdown suppers
    </Text>
    <Text
      style={
        Array [
          Object {
            "flexShrink": 0,
            "fontFamily": "GuardianTextSans-Regular",
            "fontSize": 15,
            "lineHeight": 18,
          },
          Object {
            "color": "#333333",
          },
          Object {
            "color": "#121212",
            "flexWrap": "wrap",
            "paddingRight": 20,
          },
        ]
      }
    >
      Available until 
      2/1/1998
    </Text>
  </View>
</View>
`;

exports[`SpecialEditionButton should display a selected SpecialEditionButton despite correct styling 1`] = `
<View
  accessibilityLabel="Food
Monthly - Edition Button"
  accessibilityRole="button"
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "#052962",
      "borderRadius": 3,
      "borderWidth": 4,
      "flexDirection": "row",
      "paddingBottom": 32,
      "paddingTop": 14,
    }
  }
>
  <View
    style={
      Object {
        "marginRight": 13,
        "width": 75,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "flexShrink": 0,
            "fontFamily": "GTGuardianTitlepiece-Bold",
            "fontSize": 30,
            "lineHeight": 33,
          },
          Object {
            "color": "#FEEEF7",
            "fontFamily": "GHGuardianHeadline-Medium",
            "fontSize": 32,
            "lineHeight": 34,
            "marginBottom": 5,
          },
        ]
      }
    >
      Food
Monthly
    </Text>
    <Text
      style={
        Array [
          Object {
            "flexShrink": 0,
            "fontFamily": "GuardianTextSans-Regular",
            "fontSize": 15,
            "lineHeight": 18,
          },
          Object {
            "color": "#333333",
          },
          Object {
            "color": "#121212",
            "flexWrap": "wrap",
            "paddingRight": 20,
          },
        ]
      }
    >
      Store cupboard special: 20 quick and easy lockdown suppers
    </Text>
    <Text
      style={
        Array [
          Object {
            "flexShrink": 0,
            "fontFamily": "GuardianTextSans-Regular",
            "fontSize": 15,
            "lineHeight": 18,
          },
          Object {
            "color": "#333333",
          },
          Object {
            "color": "#121212",
            "flexWrap": "wrap",
            "paddingRight": 20,
          },
        ]
      }
    >
      Available until 
      2/1/1998
    </Text>
  </View>
</View>
`;
